libbpf-tools/execsnoop.c

```diff
@@ -0,0 +1,338 @@
+        errno = 0;
+        max_args = strtol(arg, NULL, 10);
+        if (errno || max_args < 1 || max_args > TOTAL_MAX_ARGS) {
+            fprintf(stderr, "Invalid MAX_ARGS %s, should be in [1, %d] range\n",
+					arg, TOTAL_MAX_ARGS);
+
+            argp_usage(state);
+        }
+		env.max_args = max_args;
+		break;
+    default:
+        return ARGP_ERR_UNKNOWN;
+    }
+    return 0;
+}
+
+static int bump_memlock_rlimit(void)
+{
+    struct rlimit rlim_new = {
+        .rlim_cur   = RLIM_INFINITY,
+        .rlim_max   = RLIM_INFINITY,
+    };
+
+    return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+            const char *format, va_list args)
+{
+    if (level == LIBBPF_DEBUG && !env.verbose)
+        return 0;
+    return vfprintf(stderr, format, args);
+}
+
+static void time_since_start(char *str, size_t size)
+{
+	int ret;
+	long nsec, sec;
+	static struct timespec cur_time;
+
+    clock_gettime(CLOCK_MONOTONIC, &cur_time);
+	nsec = cur_time.tv_nsec - start_time.tv_nsec;
+	sec = cur_time.tv_sec - start_time.tv_sec;
+	if (nsec < 0) {
+		nsec += NSEC_PER_SEC;
+		sec--;
+	}
+	/* rounded division */
+	nsec = (nsec + (NSEC_PRECISION / 2)) / NSEC_PRECISION;
+	ret = snprintf(str, size, "%ld.%ld", sec, nsec);
+	if (ret < 0) {
+		str[size - 1] = '\0';
+	}
+}
+
+static void print_args(const struct event *e, bool quote)
+{
+	for (int i = 0; i < e->args_count && i < TOTAL_MAX_ARGS; i++) {
+        const char *c = &(e->args[i * ARGSIZE]);
+        int j = 0;
+		if (i > 0) {
+			putchar(' ');
+		}
+		/* arguments should be quoted */
+		if (env.quote) {
+			putchar('"');
+			while(*c && j < ARGSIZE) {
+				switch(*c) {
+					case '"':
+						putchar('\\');
+						putchar('"');
+						break;
+					case '\t':
+						putchar('\\');
+						putchar('t');
+						break;
+					case '\n':
+						putchar('\\');
+						putchar('n');
+						break;
+					default:
+						putchar(*c);
+						break;
+				}
+				c++;
+				j++;
+			}
+			putchar('"');
+		} else {
+			/* output arguments as is */
+			while(*c && j < ARGSIZE) {
+				putchar(*c++);
+				j++;
+			}
+		}
+    }
+    if (e->more_args) {
+		fputs(" ...", stdout);
+    }
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+    const struct event *e = data;
+    time_t t;
+    struct tm *tm;
+    char ts[32];
+	char time_since[16];
+
+	/* TODO: use pcre lib */
+	if (env.name && strstr(e->comm, env.name) == NULL)
+		return;
+
+	/* TODO: use pcre lib */
+	if (env.line && strstr(e->comm, env.line) == NULL)
+		return;
+
+    time(&t);
+    tm = localtime(&t);
+    strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+
+    if (env.time) {
+        printf("%-8s ", ts);
+    }
+	if (env.timestamp) {
+		time_since_start(time_since, sizeof(time_since));
+		printf("%-8s ", time_since);
+	}
+
+    if (env.print_uid)
+        printf("%-6d", e->uid);
+
+    printf("%-16s %-6d %-6d %3d ", e->comm, e->pid, e->ppid, e->retval);
+	print_args(e, env.quote);
```

Comment:
Sorry, for inconsistency with tabs and spaces, at first I started with spaces and then when noticed that all other tools in `libbpf-tools` folder using tabs reconfigured editor to tabs.
I'll **fix this** with next update.

Reply:
sounds good, let's keep formatting nice and consistent